File: gateway/internal/auth/keys.go
// Package auth issues, hashes, looks up, and gates HTTP requests by API key.
package auth

import (
	"crypto/rand"
	"crypto/sha256"
	"crypto/subtle"
	"encoding/base32"
	"fmt"
)

// PrefixLen is the visible portion of an issued key (e.g. "cru_live_A3F9NK4M7QHGBVTP").
// Customers see this in the dashboard; it's indexed in Postgres for O(1) lookup before
// the constant-time hash compare gates auth.
//
// With the literal "cru_live_" prefix (9 chars) eating the first slots, PrefixLen=24 leaves
// 15 base32 chars of entropy = 32^15 ≈ 3.5e22 combinations. The active-prefix index is
// UNIQUE, so we'd collide at birthday-paradox ~1e11 keys — billions of years of issuance.
// PrefixLen used to be 12 (only 3 random chars; collisions likely at hundreds of keys).
const PrefixLen = 24

// Generate returns (full_key, displayable_prefix). The full key is shown to the customer
// ONCE at creation; only the prefix and Hash(salt, full_key) are persisted.
//
// keyPrefix is the product-customizable identifier (e.g. "cru_" — set per product clone via API_KEY_PREFIX).
func Generate(keyPrefix string) (full, prefix string, err error) {
	raw := make([]byte, 24) // 192 bits of entropy
	if _, err := rand.Read(raw); err != nil {
		return "", "", fmt.Errorf("read random: %w", err)
	}
	suffix := base32.StdEncoding.WithPadding(base32.NoPadding).EncodeToString(raw)
	full = keyPrefix + "live_" + suffix

	prefix = full
	if len(prefix) > PrefixLen {
		prefix = prefix[:PrefixLen]
	}
	return full, prefix, nil
}

// Hash returns SHA-256(salt || key). Stored as bytea in api_keys.hash.
func Hash(salt, key string) []byte {
	h := sha256.New()
	h.Write([]byte(salt))
	h.Write([]byte(key))
	return h.Sum(nil)
}

// VerifyHash compares two hashes in constant time.
func VerifyHash(a, b []byte) bool {
	return subtle.ConstantTimeCompare(a, b) == 1
}

File: dashboard/lib/keys.ts
// API key generation and hashing — MUST match gateway/internal/auth/keys.go on the byte level.
// The gateway and dashboard read/write the same api_keys table; if their hash functions diverge,
// dashboard-issued keys will not authenticate.
// NOTE: Validation for API key names lives in dashboard/app/api/keys/route.ts
import { randomBytes, createHash } from "crypto";

// MUST match auth.PrefixLen in gateway/internal/auth/keys.go.
const PREFIX_LEN = 24;
const BASE32_ALPHABET = "ABCDEFGHIJKLMNOPQRSTUVWXYZ234567";

/**
 * Generate a new API key. Returns the full key (shown to the customer once)
 * and the displayable prefix (stored alongside the hash for O(1) lookup).
 */
export function generateKey(productPrefix: string): { full: string; prefix: string } {
  const raw = randomBytes(24); // 192 bits, matches Go's auth.Generate
  const suffix = encodeBase32(raw);
  const full = `${productPrefix}live_${suffix}`;
  const prefix = full.slice(0, PREFIX_LEN);
  return { full, prefix };
}

/**
 * SHA-256(salt || key). Must match gateway/internal/auth/keys.go:Hash() byte-for-byte.
 */
export function hashKey(salt: string, key: string): Buffer {
  const h = createHash("sha256");
  h.update(salt);
  h.update(key);
  return h.digest();
}

// Standard base32 (RFC 4648), no padding — same alphabet Go's encoding/base32 uses.
function encodeBase32(bytes: Buffer): string {
  let bits = 0;
  let value = 0;
  let output = "";
  for (const byte of bytes) {
    value = (value << 8) | byte;
    bits += 8;
    while (bits >= 5) {
      output += BASE32_ALPHABET[(value >>> (bits - 5)) & 31];
      bits -= 5;
    }
  }
  if (bits > 0) {
    output += BASE32_ALPHABET[(value << (5 - bits)) & 31];
  }
  return output;
}
